File: 223_Rectangle_Area.java
/*
https://leetcode.com/problems/rectangle-area/
Find the total area covered by two rectilinear rectangles in a 2D plane.
Each rectangle is defined by its bottom left corner and top right corner.

Calculate left, right, top, bottom edges of overlap.  If inverted then no overlap, else subtract
overlap area from sum of rectangles.
Time - O(1)
Space - O(1)
 */

public class Solution {
    public int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {

        int right_edge = Math.min(C, G);
        int left_edge = Math.max(A, E);

        int top_edge = Math.min(D, H);
        int bottom_edge = Math.max(B, F);

        int overlap = 0;
        if (right_edge > left_edge && top_edge > bottom_edge)
            overlap = (right_edge - left_edge) * (top_edge - bottom_edge);

        int r1 = (C-A) * (D-B);
        int r2 = (G-E) * (H-F);

        return r1 + r2 - overlap;
    }
}

File: 223_Rectangle_Area.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/rectangle-area/
# Find the total area covered by two rectilinear rectangles in a 2D plane.
# Each rectangle is defined by its bottom left corner (A, B) or (E, F) and top right corner (C, D) or (G, H).

# Calculate the left and right edges of horizontal overlap. Take the difference to get the x_overlap and similar for
# y_overlap. Subtract the overlap area from the sum of the areas of the two rectangles.
# Time - O(1)
# Space - O(1)

class Solution(object):
    def computeArea(self, A, B, C, D, E, F, G, H):
        """
        :type A: int
        :type B: int
        :type C: int
        :type D: int
        :type E: int
        :type F: int
        :type G: int
        :type H: int
        :rtype: int
        """
        x_lhs = max(A, E)
        x_rhs = min(C, G)
        x_overlap = max(x_rhs - x_lhs, 0)

        y_lhs = max(B, F)
        y_rhs = min(D, H)
        y_overlap = max(y_rhs - y_lhs, 0)

        rect1 = (C - A) * (D - B)
        rect2 = (G - E) * (H - F)

        return rect1 + rect2 - y_overlap * x_overlap